core/amber/src/main/resources/python_udf/texera_udf_main.py

```diff
@@ -1,65 +1,85 @@
-                     file_log_dir: str, file_log_level: str, file_log_fmt: str, file_datefmt: str) -> None:
+class InterceptHandler(logging.Handler):
+    def emit(self, record):
+        # Get corresponding Loguru level if it exists
+        try:
+            level = logger.level(record.levelname).name
+        except ValueError:
+            level = record.levelno
+
+        # Find caller from where originated the logged message
+        frame, depth = logging.currentframe(), 2
+        while frame.f_code.co_filename == logging.__file__:
+            frame = frame.f_back
+            depth += 1
+
+        logger.opt(depth=depth, exception=record.exc_info).log(level, record.getMessage())
+
+
+class StreamToLogger(object):
+
+    def __init__(self, level="INFO"):
+        self._level = level
+
+    def write(self, buffer):
+        for line in buffer.rstrip().splitlines():
+            logger.opt(depth=1).log(self._level, "[print] " + line.rstrip())
+
+    def flush(self):
+        pass
+
+
+def init_root_logger(stream_log_level: str, stream_log_fmt: str,
+                     file_log_dir: str, file_log_level: str, file_log_fmt: str) -> None:
     """
     initialize root logger with the given configurations
-    :param log_input_level: level to be captured by the root logger
     :param stream_log_level: level to be output to stdout/stderr
     :param stream_log_fmt: log format to stdout/stderr
-    :param stream_datefmt: log datefmt to stdout/stderr
     :param file_log_dir: log file directory
     :param file_log_level: level to be output to file
-    :param file_log_fmt: log format to file
-    :param file_datefmt: log datefmt to file
+    :param file_log_fmt: log format to fil
     :return:
     """
-
-    logger = logging.getLogger()
-    logger.setLevel(log_input_level)
-
+    logger.remove()
     # set up stream handler, which outputs to stdout and stderr
-    stream_handler = logging.StreamHandler()
-    stream_handler.setLevel(stream_log_level)
-    stream_handler.setFormatter(logging.Formatter(stream_log_fmt, datefmt=stream_datefmt))
-    logger.addHandler(stream_handler)
+    logger.add(sys.stderr, format=stream_log_fmt, level=stream_log_level)
 
     # set up file handler, which outputs log file
     file_name = f"texera-python_udf-{datetime.utcnow().isoformat()}-{os.getpid()}.log"
     file_path = Path(file_log_dir).joinpath(file_name)
-    file_handler = logging.FileHandler(file_path)
-
-    file_handler.setLevel(file_log_level)
-    file_handler.setFormatter(logging.Formatter(file_log_fmt, datefmt=file_datefmt))
     logger.info(f"Attaching a FileHandler to logger, file path: {file_path}")
-    logger.addHandler(file_handler)
+    logger.add(file_name, format=file_log_fmt, level=file_log_level)
     logger.info(f"Logger FileHandler is now attached, previous logs are in StreamHandler only.")
 
+    # intercept standard logging module to loguru
+    logging.basicConfig(handlers=[InterceptHandler()], level=0)
```

Comment:
In general this file is very heavy. I suggest breaking it into smaller files. Also to a person unfamiliar with python, it's not clear how you are doing the logging. Please put detailed comments to explain in lay man's terms

Reply:
I agree this file is getting bigger. It contains three parts:
1. configure logger.
2. register/load/link the udf as a module.
3. launch the PyArrow server.


We could do a refactoring PR to break them into much smaller pieces. For this PR, I suggest we keep it for now.